### PR TITLE
add title attribute to form

### PR DESCRIPTION
**Pull request by mhilton**

### CODE REVIEW

File: httpbakery/form/form.go [hunk 1]
```diff
@@ -49,11 +50,17 @@ client's responsibility to interpret the schema and present it to the
 user.
 */
 
-// Filler represents an object that can fill out a form. The given schema
-// represents the form template. The returned value should be compatible
-// with this.
+// Form describes a form that needs to be completed to perform the login.
+type Form struct {
```
Comment (rogpeppe) [inline httpbakery/form/form.go]: +1 to making this a struct, as I can definitely see the potential for more fields here in the future.

File: httpbakery/form/form.go [hunk 2]
```diff
@@ -49,11 +50,17 @@ client's responsibility to interpret the schema and present it to the
...
+	Title  string
+	Fields environschema.Fields
+}
+
+// Filler represents an object that can fill out a form. The the form is
+// described in f. The returned value should be compatible with the
+// schema defined in f.Fields.
 type Filler interface {
```
Comment (rogpeppe) [inline httpbakery/form/form.go]: It crossed my mind that we should be using environschema/form.Filler before, but now I'm more sure of that.

File: httpbakery/form/form.go [hunk 3]
```diff
@@ -163,7 +170,14 @@ func (v webPageVisitor) visitWebPage(u *url.URL) error {
 	if len(s.Schema) == 0 {
 		return errgo.Newf("invalid schema: no fields found")
 	}
-	form, err := v.filler.Fill(s.Schema)
+	host, err := publicsuffix.EffectiveTLDPlusOne(u.Host)
+	if err != nil {
+		host = u.Host
+	}
+	form, err := v.filler.Fill(Form{
+		Title:  "Login to " + host,
```
Comment (rogpeppe) [inline httpbakery/form/form.go]: s/Login/Log in/

File: httpbakery/form/form_test.go [hunk 4]
```diff
@@ -256,12 +306,53 @@ var testError = &httpbakery.Error{
...
+type titleTestFiller struct {
+	title string
+}
+
+func (f *titleTestFiller) Fill(frm form.Form) (map[string]interface{}, error) {
+	f.title = frm.Title
+	return map[string]interface{}{"test": 1}, nil
+}
+
+type transport struct {
```
Comment (rogpeppe) [inline httpbakery/form/form_test.go]: A doc comment here would be useful - I spent a little time trying to work out what this was actually doing.

// transport implements an HTTP transport that wraps rt
// and replaces the specified prefix string in outgoing request
// addresses with the specified address.

?

File: httpbakery/form/form_test.go [hunk 5]
```diff
@@ -256,12 +307,30 @@ var testError = &httpbakery.Error{
...
+
+func (l testLocator) PublicKeyForLocation(loc string) (*bakery.PublicKey, error) {
+	return l.locator.PublicKeyForLocation(l.loc)
+}
+
+type titleTestFiller struct {
+	title string
+}
+
+func (f *titleTestFiller) Fill(frm esform.Form) (map[string]interface{}, error) {
```
Comment (rogpeppe) [inline httpbakery/form/form_test.go]: s/frm/form/ ?
(as who cares if we shadow the form package identifier?)

File: bakerytest/bakerytest.go [hunk 6]
```diff
@@ -221,8 +222,9 @@ func (d *InteractiveDischarger) checker(req *http.Request, cavId, cav string) ([
 	d.id++
 	d.waiting[id] = discharge{cavId, make(chan dischargeResult, 1)}
 	d.mu.Unlock()
-	visitURL := fmt.Sprintf("%s/visit?waitid=%s", d.Discharger.server.URL, id)
-	waitURL := fmt.Sprintf("%s/wait?waitid=%s", d.Discharger.server.URL, id)
+	prefix := strings.TrimSuffix(req.URL.String(), "/discharge")
```
Comment (rogpeppe) [inline bakerytest/bakerytest.go:225]: Why has this changed? What was wrong with d.Discharger.server.URL ?

**Comment by rogpeppe** [comment]
LGTM, thanks very much.

File: bakerytest/bakerytest.go (hunk 6 above)
Comment (mhilton) [inline bakerytest/bakerytest.go:225]: The title of the form is derived from the host part of the URL. I have made this change to make the host part of the URL predictable.

File: httpbakery/form/form_test.go (hunk 5 above)
Comment (mhilton) [inline httpbakery/form/form_test.go]: done

File: httpbakery/form/form_test.go [hunk 7]
```diff
@@ -159,6 +161,55 @@ func (s *formSuite) TestFormLoginNewRequestError(c *gc.C) {
 	c.Assert(err, gc.ErrorMatches, "cannot create request: parse :://: missing protocol scheme")
 }
 
+var formTitleTests = []struct {
```
Comment (mattyw) [inline httpbakery/form/form_test.go:164]: Why not do these inline in TestFormTitle?

**Comment by mattyw** [comment]
LGTM  - one nitpick for consideration

File: httpbakery/form/form_test.go (hunk 7 above)
Comment (mhilton) [inline httpbakery/form/form_test.go:164]: Its generally easier to add things to tables.

**Pull request merged**